src/mongocrypt-ctx.c

```diff
@@ -821,4 +824,99 @@ mongocrypt_ctx_setopt_masterkey_aws_endpoint (mongocrypt_ctx_t *ctx,
    }
    ctx->opts.masterkey_aws_endpoint_len = endpoint_len;
    return true;
+}
+
+bool
+mongocrypt_ctx_setopt_key_encryption_key (mongocrypt_ctx_t *ctx,
+                                          mongocrypt_binary_t *bin)
+{
+   bson_t as_bson;
+   bson_iter_t iter;
+   const char *value;
+   uint32_t len;
+
+   if (!ctx) {
+      return false;
+   }
+
+   if (ctx->initialized) {
+      return _mongocrypt_ctx_fail_w_msg (ctx, "cannot set options after init");
+   }
+
+   if (ctx->state == MONGOCRYPT_CTX_ERROR) {
+      return false;
+   }
+
+   if (ctx->opts.masterkey_kms_provider) {
+      return _mongocrypt_ctx_fail_w_msg (ctx, "key encryption key already set");
+   }
+
+   if (!_mongocrypt_binary_to_bson (bin, &as_bson)) {
+      return _mongocrypt_ctx_fail_w_msg (ctx, "invalid BSON");
+   }
+
+   if (!bson_iter_init_find (&iter, &as_bson, "provider") ||
+       !BSON_ITER_HOLDS_UTF8 (&iter)) {
+      return _mongocrypt_ctx_fail_w_msg (ctx, "expected UTF-8 provider");
+   }
+
+   if (0 != strcmp (bson_iter_utf8 (&iter, NULL), "azure")) {
+      /* TODO: support GCP, Azure, and Local through this API */
+      return _mongocrypt_ctx_fail_w_msg (ctx, "expected 'azure' for provider");
+   }
+
+   if (!bson_iter_init_find (&iter, &as_bson, "keyVaultEndpoint") ||
+       !BSON_ITER_HOLDS_UTF8 (&iter)) {
+      return _mongocrypt_ctx_fail_w_msg (ctx,
+                                         "expected UTF-8 keyVaultEndpoint");
+   }
+
+   value = bson_iter_utf8 (&iter, &len);
+   ctx->opts.azure_kek.key_vault_endpoint =
+      _mongocrypt_endpoint_new (value, len, ctx->status);
+   if (!ctx->opts.azure_kek.key_vault_endpoint) {
+      return false;
+   }
+
+   if (!bson_iter_init_find (&iter, &as_bson, "keyName") ||
+       !BSON_ITER_HOLDS_UTF8 (&iter)) {
+      return _mongocrypt_ctx_fail_w_msg (ctx, "expected UTF-8 keyName");
+   }
+
+   value = bson_iter_utf8 (&iter, &len);
+   if (!_mongocrypt_validate_and_copy_string (
+          value, len, &ctx->opts.azure_kek.key_name)) {
+      return _mongocrypt_ctx_fail_w_msg (ctx, "invalid keyName");
+   }
+
+   if (bson_iter_init_find (&iter, &as_bson, "keyVersion")) {
+      if (!BSON_ITER_HOLDS_UTF8 (&iter)) {
```

Comment:
Since this is optional, would `BSON_ITER_HOLDS_NULL` be ok, or would `keyVersion` (the key itself) need to be absent?

Reply:
I don't mind either way.